src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/FindAllReferencesHandler.cs

```diff
@@ -78,35 +95,85 @@ public async Task<VSReferenceItem[]> HandleRequestAsync(ReferenceParams request,
 
             cancellationToken.ThrowIfCancellationRequested();
 
-            var serverKind = LanguageServerKind.CSharp;
-            var referenceParams = new ReferenceParams()
+            // Temporary till IProgress serialization is fixed
+            var requestId = Guid.NewGuid().ToString(); // request.PartialResultToken.Id
+
+            var referenceParams = new CustomReferenceParams()
             {
                 Position = projectionResult.Position,
                 TextDocument = new TextDocumentIdentifier()
                 {
                     Uri = projectionResult.Uri
                 },
                 Context = request.Context,
-                PartialResultToken = request.PartialResultToken
+                PartialResultToken = requestId // request.PartialResultToken
             };
 
-            var result = await _requestInvoker.ReinvokeRequestOnServerAsync<ReferenceParams, VSReferenceItem[]>(
-                Methods.TextDocumentReferencesName,
-                serverKind,
-                referenceParams,
-                cancellationToken).ConfigureAwait(false);
+            var waitForProgressCompletion = new ManualResetEvent(false);
+
+            Task callback(JToken value)
+            {
+                return ProcessReferenceItemsAsync(value, request.PartialResultToken, cancellationToken);
+            }
+
+            var callbackRequest = new CallbackRequest(
+                callback,
+                WaitForProgressNotificationTimeout,
+                waitForProgressCompletion);
+
+            try
+            {
+                var isSubscribed = _lspProgressListener.Subscribe(callbackRequest, requestId /* IProgress<T> */);
+                if (!isSubscribed)
+                {
+                    return null;
+                }
+
+                _ = await _requestInvoker.ReinvokeRequestOnServerAsync<CustomReferenceParams, VSReferenceItem[]>(
+                    Methods.TextDocumentReferencesName,
+                    LanguageServerKind.CSharp,
+                    referenceParams,
+                    cancellationToken).ConfigureAwait(false);
+
+                cancellationToken.ThrowIfCancellationRequested();
 
+                // We must not return till we have received the progress notifications
+                // and reported the results via the PartialResultToken
+                waitForProgressCompletion.WaitOne(WaitForProgressCompletionTimeout);
```

Comment:
This blocks the current thread which I don't think we want. I'd suggest changing this design slightly. What if the progress subscription (CallbackRequest) exposed an "OnComplete" task (under the covers it'd probably use a task completion source) where once the task was completed it'd automatically "unsubscribe" from the active requests list? That way this code would look more like:

```C#
if (!_lspProgressListener.TryListenForProgress(requestId,
	onProgressResult: () => ProcessReferenceItemsAsync(value, request.PartialResultToken, cancellationToken),
	out var onCompleted)
{
    return null;
}

_ = await _requestInvoker.ReinvokeRequestOnServerAsync<CustomReferenceParams, VSReferenceItem[]>(
	Methods.TextDocumentReferencesName,
	LanguageServerKind.CSharp,
	referenceParams,
	cancellationToken).ConfigureAwait(false);

cancellationToken.ThrowIfCancellationRequested();

await onCompleted;
```

The benefit of async/await allows the currently executing thread to stop processing the current task and move to other tasks; however, when you do "WaitOne" it blocks the thread synchronously and doesn't allow the thread to process any other info.

Reply:
Thanks for the feedback, that makes a lot of sense! I've updated accordingly.